doc/decisions/memory_layout.md

```diff
@@ -0,0 +1,32 @@
+# Memory Layout
+
+## Problem
+
+`mmapstorage` leaks internals of the `struct _Key`.
+
+## Constraints
+
+## Assumptions
+
+## Considered Alternatives
+
+- don't make `Key` opaque
+- don't provide `mmapstorage`
+
+## Decision
+
+Have versioning and memory layout checks of cache files written by `mmapstorage`.
```

Comment:
Unless I'm missing something important, this cannot work.

1. `mmapstorage` is written C, so it itself can only create `Key`s/`KeySet`s using the memory layout of the C implementation.
2. The magic numbers `mmapstorage` uses only guard against byte-order changes, e.g. if the cache was copied between machines with different architectures.
3. If you give `mmapstorage` a `KeySet *` that uses a different memory layout, it will happily try and interpret that as a `struct _KeySet *` using the C memory layout. With a lot of luck it might even succeed in storing _something_ into the cache.
4. If a cache file was written by `mmapstorage`, it will be able to read it on the same architecture. It will happily treat the data from the cache as is if it followed the C memory layout. That means it will read it's own magic data and verify that is correct. This will work, because `mmapstorage` always uses the C memory layout. Then it is going to update the pointers. Or at least it pretends to, because where `mmapstorage` thinks the pointers are is probably not where the pointers are in the completely different memory layout. Finally, it will assign some pointer to where it thinks the KeySet array is, but probably it will just write garbage to a random memory location, or even segfault.

The only way this can work AFAIK is one of:
1. **ALL** implementation use the exact same memory layout (down to the order of bits in the flags field) for `Key`  and `KeySet`
2. There is a separate `mmapstorage` implementation for every memory layout. Alternatively, there could be a `mmap` support library for every memory layout that holds all the functions that assume a certain memory layout

There would be a bit more flexibility, if we didn't use pointers, but that is not really an option.